### PR TITLE
Support `Path.addCircle`

### PR DESCRIPTION
**Pull request by dnfield**
For some reason this is missing currently (can be achieved with `addOval`, which is implemented in dart and which is how Skia implements it anyway on the C++ side).

I'd like to see this to support SVG rendering (particularly, making circles into paths).

See also https://github.com/flutter/flutter/issues/15501

### CODE REVIEW

**Comment by Hixie** [comment]
This seems somewhat redundant with `arcTo`; is `arcTo` not sufficient?

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -1519,6 +1519,13 @@ class Path extends NativeFieldWrapperClass2 {
   }
   void _addOval(double left, double top, double right, double bottom) native 'Path_addOval';
 
+  /// Adds a new subpath that consists of a curve that forms the
+  /// circle centered on centerX,centerY with specified radius
```
Comment (Hixie) [inline lib/ui/painting.dart]: This should end with a period. Also, put backticks around references to arguments.

**Comment by Hixie** [comment]
If we add this, it would be nice if we could also find a way to test it. (I don't think we yet test much of this code, but we really should.)

**Comment by dnfield** [comment]
The following code all would result in the same end result in terms of drawing:
```
const cx = 20.0;
const cy = 20.0;
const r = 10.0;
Path c1 = new Path()..addCircle(cx, cy, r);
Path c2 = new Path()
  ..addOval(new Rect.fromLTRB(cx - r, cy - r, cx + r, cy + r));
Path c3 = new Path()
  ..addArc(new Rect.fromLTRB(cx - r, cy - r, cx + r, cy + r), 0.0, 360.0);
```

But `addCircle` is the clearest/most idiomatic way of doing it (rather than needing to do some math to realize the 'Oval' or 'Arc' you're adding is really a circle).

The dart code in painting.dart could be rewritten to be an alias to one of those (which is what Skia does in C++ - `SkPath::addCircle` calls `SkPath::addOval` using the formula above).

**Comment by dnfield** [comment]
RE Testing: there are some tests for painting in flutter/flutter, but nothing around paths.  I'm not really sure how we could test Paths as they are today - there's no way to test for value equality or get any insight into the contents of a path right now.

**Comment by Hixie** [comment]
I was thinking more:
```dart
Path c4 = new Path()..arcTo(new Rect.fromCircle(center: new Offset(cx, cy), radius: r), 0, pi * 2.0);
```
...but as you say, there's at least two other ways to do it too. Do we really want to add a fourth?

**Comment by Hixie** [comment]
(Does the existence of `Rect.fromCircle` affect your opinion, by the way?)

**Comment by dnfield** [comment]
Yes, `arcTo` is fine - although I believe `addCircle` includes an implicit `moveTo`.  

I don't see this as a particularly critical change - there are numerous work arounds.  However, I think it will be an easy source of confusion for developers (particularly when trying to port existing path based logic from other languages/contexts).  I think at the least it would make sense to support `addCircle` like so in `painting.dart`:

```
void addCircle(double centerX, double centerY, double radius) {
  _addOval(new Rect.fromCircle(center: new Offset(centerX, centerY), r: radius));
}
```

I realize this is adding yet another way to achieve the same thing, but the existing alternatives are less immediately obvious if you just want to add a circle to a path and prone to introducing errors/increased need for unit testing.

**Comment by Hixie** [comment]
I would definitely support adding documentation to `addOval` and `drawOval` that talks about how if you want to draw a circle from a radius and a center point, you can use `Rect.fromCircle` to do it easily, but I don't see much value in adding more API surface. Keeping the engine API surface small is a goal.

**Comment by dnfield** [comment]
Ok.  It may make sense to just create a Path drawing package in dartpub as well with "utility" methods like these.  That should be entirely possible with no engine modifications.

Would you like me to modify this PR to update documentation on `addOval` as suggested?

**Comment by Hixie** [comment]
Sure. Thanks!

**Comment by dnfield** [comment]
I only updated Path.addOval - canvas already has an exposed method `drawCircle`.

**Comment by Hixie** [comment]
LGTM

**Comment by Hixie** [comment]
When you said Canvas already had "drawCircle" I was like, wait, what? When did we add that. That's silly.
So I did some digging.
Turns out that drawCircle was in fact the first thing we added. The first version of the canvas API literally supporting nothing _but_ drawCircle. -_-

https://github.com/flutter/engine/blame/1eb17996fbac62d075b8805f7fef7ac7ae551897/engine/core/painting/Canvas.idl

**Pull request merged**